### PR TITLE
Model fitting residuals

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request implements the option to expose the residuals from a model fit as an additional entry in the data collection.

Open questions:
* should the residuals be automatically added to the viewer (if the switch to add the model is checked), should it have an independent switch, or should it not be added (current behavior)?

https://user-images.githubusercontent.com/877591/203401264-dfa904e6-a1ed-4a41-b66f-33bd6c8c8d2f.mov

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:803]: It seems that during past refactors this kind of got left as an unnecessary artifact that only gets called once internally (and so never even entered the else statement in 797).  This PR moves the remaining applicable logic to where it was called before.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.vue
Comment (rosteen) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.vue:201]:
```suggestion
              hint="Whether to compute and export residuals (data - model)."
```
Is this perhaps more readable as "data minus model"?

Comment (kecnry) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.vue:201]: I like that even more (to avoid confusion with hyphenation), thanks!

**Comment by rosteen** [comment]
Works as expected in Specviz and Cubeviz, I'll give the code quick look-over in the morning before I approve.

**Comment by camipacifici** [comment]
Looks lovely! But I hit a traceback if I have two spectra loaded, I highlight a subset, and then I select it in the model fitting plugin.
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/ipywidgets/widgets/widget.py:756, in Widget._handle_msg(self, msg)
    754         if 'buffer_paths' in data:
    755             _put_buffers(state, data['buffer_paths'], msg['buffers'])
--> 756         self.set_state(state)
    758 # Handle a state request.
    759 elif method == 'request_state':

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/ipywidgets/widgets/widget.py:620, in Widget.set_state(self, sync_data)
    615         self._send(msg, buffers=echo_buffers)
    617 # The order of these context managers is important. Properties must
    618 # be locked when the hold_trait_notification context manager is
    619 # released and notifications are fired.
--> 620 with self._lock_property(**sync_data), self.hold_trait_notifications():
    621     for name in sync_data:
    622         if name in self.keys:

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/contextlib.py:142, in _GeneratorContextManager.__exit__(self, typ, value, traceback)
    140 if typ is None:
    141     try:
--> 142         next(self.gen)
    143     except StopIteration:
    144         return False

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/traitlets/traitlets.py:1371, in HasTraits.hold_trait_notifications(self)
   1369 for changes in cache.values():
   1370     for change in changes:
-> 1371         self.notify_change(change)

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/ipywidgets/widgets/widget.py:686, in Widget.notify_change(self, change)
    683     if name in self.keys and self._should_send_property(name, getattr(self, name)):
    684         # Send new state to front-end
    685         self.send_state(key=name)
--> 686 super(Widget, self).notify_change(change)

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/traitlets/traitlets.py:1386, in HasTraits.notify_change(self, change)
   1384 def notify_change(self, change):
   1385     """Notify observers of a change event"""
-> 1386     return self._notify_observers(change)

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/traitlets/traitlets.py:1431, in HasTraits._notify_observers(self, event)
   1428 elif isinstance(c, EventHandler) and c.name is not None:
   1429     c = getattr(self, c.name)
-> 1431 c(event)

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/jdaviz/core/template_mixin.py:874, in SubsetSelect._selected_changed(self, event)
    872 def _selected_changed(self, event):
    873     super()._selected_changed(event)
--> 874     self._update_has_subregions()

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/jdaviz/core/template_mixin.py:925, in SubsetSelect._update_has_subregions(self)
    923     self.selected_has_subregions = False
    924 else:
--> 925     self.selected_has_subregions = len(self.selected_obj.subregions) > 1

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/functools.py:981, in cached_property.__get__(self, instance, owner)
    979 val = cache.get(self.attrname, _NOT_FOUND)
    980 if val is _NOT_FOUND:
--> 981     val = self.func(instance)
    982     try:
    983         cache[self.attrname] = val

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/jdaviz/core/template_mixin.py:936, in SubsetSelect.selected_obj(self)
    932 # NOTE: we use reference names here instead of IDs since get_subsets_from_viewer requires
    933 # that.  For imviz, this will mean we won't be able to loop through each of the viewers,
    934 # but the original viewer should have access to all the subsets.
    935 for viewer_ref in self.viewer_refs:
--> 936     match = self.app.get_subsets_from_viewer(viewer_ref,
    937                                              subset_type=subset_type).get(self.selected)
    938     if match is not None:
    939         return match

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/jdaviz/app.py:815, in Application.get_subsets_from_viewer(self, viewer_reference, data_label, subset_type)
    812 else:
    813     data_wcs = value.data.coords
--> 815 subregions_in_subset = _get_all_subregions(
    816         np.where(value.to_mask() == True)[0], # noqa
    817         data_wcs.spectral_axis)
    819 regions[key] = subregions_in_subset
    820 continue

File ~/opt/miniconda3/envs/pr1864/lib/python3.10/site-packages/jdaviz/app.py:742, in Application.get_subsets_from_viewer.<locals>._get_all_subregions(mask, spec_axis_data)
    725 """
    726 Return all subregions within a subset.
    727 
   (...)
    737     SpectralRegion object containing all subregions of the subset.
    738 """
    739 if len(mask) == 0:
    740     # Mask should only be 0 if ApplyROI is used to incorrectly
    741     # create subsets via the API
--> 742     raise ValueError("Mask has length 0, ApplyROI may have been used incorrectly")
    744 current_edge = 0
    745 combined_spec_region = None

ValueError: Mask has length 0, ApplyROI may have been used incorrectly

**Comment by rosteen** [comment]
Interesting - I didn't hit that error with multiple loaded spectra (the top two in the screenshot), but maybe I didn't replicate your workflow exactly.

Screen Shot 2022-11-22 at 5 05 16 PM

**Comment by camipacifici** [comment]
The solution for the Mosviz data dropdown backfired :( the residual is not appearing in the list
Screen Shot 2022-11-22 at 9 37 53 PM

**Comment by kecnry** [comment]
@camipacifici - I wouldn't expect that bug was introduced here, but can you either provide more details to reproduce or test on main and confirm if it happens there as well or not?  As for mosviz, that's a good catch, that'll definitely need to be fixed before merging this, so I'll mark this back as draft until I get a chance to fix assigning the mosviz row/plugin information.

**Comment by camipacifici** [comment]
@kecnry I confirm that the bug is in main too. I will create a ticket for it. Thanks!

**Comment by kecnry** [comment]
I think mosviz is now working correctly (although note that because of the small size of the viewers and the data menu now being attached in #1712, some resizing might be necessary in order to access the additional entries).

One slightly annoying behavior is that the residuals entry is not added to the viewer by default (as consistent with other viztools in this PR), but will be loaded by default on a row change.  Should residuals be plotted by default for all viztools instead?

**Comment by camipacifici** [comment]
Looks great (disco mode included!)
I don't think I have an answer to your question. I will have to use it with a science case in mind and see what is annoying and what is nice to have. I am happy with it as is.

**Review by rosteen — APPROVED** [review]
LGTM, thanks.

**Review by javerbukh — APPROVED** [review]
Looks good! I did notice the mosviz behavior with row switching but if you're ok with that being follow-up work thats fine.

**Comment by kecnry** [comment]
> I did notice the mosviz behavior with row switching but if you're ok with that being follow-up work thats fine.

I'd be happy to make the changes here if we knew the behavior we want... but I suspect we might need to use it a little while first to get a feel for the expected behavior (in which case deferring probably makes sense).  I'll leave this open for a little while in case anyone has thoughts on that and will merge if I don't here anything by Monday will merge as-is and we can revise down the road if necessary.

**Pull request merged**
